### ISSUE TITLE
Framework: Filter bar height structure is wonky

### ISSUE THREAD

**Issue by hoverduck** [Automattic/wp-calypso#109]
The current implementation that manages the height of the filter bar is a bit wonky.

![image](https://cloud.githubusercontent.com/assets/7233112/11248157/a392ce8e-8ded-11e5-973d-216351d3f30e.png)

![image](https://cloud.githubusercontent.com/assets/7233112/11248163/a9cf160e-8ded-11e5-81da-f56596c5832e.png)

Specifically, because:
1. It can contain taller elements, or not
2. The height is determined by its content
3. The height is **also** determined by breakpoint
4. The height is **also** determined by JS when there is overflow (too many items for the space available).
5. The search field height (`is-pinned is-open search`) is dependent on the height... of the other items in the bar, which means that they need to be still present in the field (not removed with a proper state change) creating cross-dependence.

All of this means that height isn't reliable, and there's a lot of cross-dependence between the items and configurations inside that, which leads to hard to maintain code and overall challenges in improving further the component.
## Fixes?

There might be different approaches on how to fix this, for example:
1. Fix `section-nav` (the container) with `height: 60px` all the time, and fix any bug that cascade down from that. 
2. Make sure that every component inside the filter bar, regardless of breakpoint / else, is always of hte same height. This is similar to the above, but without fixing a specific height.
3. Other?

Adding that:
- This became apparent when getting rid of absolute positioning of the Search component within SectionNav, which was the only mechanism guaranteeing that, when SectionNav switched to mode "only show the expanded Search component", the height was preserved, based on Search's siblings.
- I also seriously considered, instead of altering `SectionNav` to better work with `Search`, creating a separate component called something like `SectionNavWithSearch`, which would be passed all the children you'd like, except for `Search`. Instead, any search-related props would be passed directly to `SectionNavWithSearch`, which would be responsible for placing a `Search` component in the appropriate place. Down the road, regardless of style decisions, that could be a more sensible interface, as opposed to the current approach of parsing children to determine which are _special_.

**Comment by lancewillett** [comment]
This one needs a Design Review to make sure things are working correctly now after recent changes.

**Comment by folletto** [comment]
The component(s) weren't refactored afaik. I still did a test for good measure since there were some changes for sure.

### How to check

I went through the various pages, and checked variations in components at different widths. Mobile/Desktop are two obvious ranges, but there are also tricky mid-points, the main one being a smaller desktop resolution that however has too many items in the bar so it collapses the tabs in a dropdown (different from the mobile one).

### Results

1. Stats — `50px desk` + `46px mobile` — tabs:✓, search:ø, counters:ø
2. Posts — `51px/59px* desk` + `46px mobile` — tabs:✓, search:✓, counters:✓, *dropdown:✓
3. Pages — `51px/59px* desk` + `46px mobile` — tabs:✓, search:✓, counters:✓, *dropdown:✓
4. Themes — `51px desk` + `46px mobile` — tabs:✓, search:✓, counters:✓, dropdown:✓
5. Domains — `50px desk` + `46px on mobile` — tabs:✓, search:ø, counters:ø, dropdown:ø
6. Settings — `51px/59px* desk` + `46px on mobile` — tabs:✓, search:ø, counters:ø, *dropdown:✓

Notes: 

* the **search** part now doesn't seem to move anymore outside the tabs/filter bar in mobile, which if I recall correctly was one of the reasons of the previous DOM architecture, which means that it can now probably be refactored as nested React components, properly.
* **Themes** seems the most stable of all, but because was custom made. Should not be used as a reference.

**tl;dr: the problem is still present and requires a refactoring.**
